[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139220
* #139208
* #139159
* #139090
* #139084
* #139081
* #138275
* #138005
* __->__ #137918
* #139083
* #139082

This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Testing: To check perf, I ran python torchchat.py generate stories110M
--dtype fp16 --device cpu on an x86 machine without AVX512FP16. Observed roughly 5x tokens/sec increase.
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

swolchok committed:
[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM

This is the first big milestone we've been building towards!
(TODO: also hook it up to GEMV in the same way fp16_gemv_trans is hooked up)

Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

jgong5 reviewed (commented):
Do you have performance numbers?

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok commented:
> Do you have performance numbers?

it improves decoding performance about 5x for `python torchchat.py generate stories110M --dtype fp16 --device cpu` on an x86 machine without AVX512FP16.

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/CPUBlas.cpp:
```diff
@@ -351,7 +352,15 @@ void gemm(
+   // is to upconvert to fp32 and call sgemm. We can do better by
+   // fusing the conversion.
+   const bool fp16_gemv_trans_fast_path_would_be_beneficial =
+     cpuinfo_initialize() && cpuinfo_has_x86_f16c() && !cpuinfo_has_x86_avx512fp16();
```
I guess checking `cpuinfo_has_x86_avx512fp16` is not necessary since onednn (mkldnn) won't use avx512fp16 to compute gemms by default because the avx512fp16 fma would incur accuracy loss.

swolchok replied:
@jgong5 https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/FP16-GEMM-using-AVX512-on-Sapphire-Rapids/m-p/1570739 doesn't seem to agree. I'm also confused -- I thought FMA was supposed to improve accuracy because it has high internal precision, so the result of the multiply doesn't have to be rounded to FP16 before the addition.

jgong5 replied:
@swolchok The link you referred to is about MKL not oneDNN (or mkldnn). MKL has dedicated API (hgemm) that uses AVX512_FP16 instruction but users should be aware of the accuracy loss due to FP16 accumulators. It is not about the accumulator for a single FMA (which has high internal precision has you mentioned) but about accumulation along the K-dim across multiple FMAs. On the other hand, oneDNN uses FP32 accumulators to keep high accuracy.

swolchok replied:
oh, we use FP32 accumulation as well.

malfet replied:
`mkldnn_fp16_gemm` despite the name is also available on ARM, it looks like your change will skip MKLDNN unless it is on x86 platform, wouldn't it? What is the motivation for it?

swolchok replied:
> mkldnn_fp16_gemm despite the name is also available on ARM,

news to me!

> it looks like your change will skip MKLDNN unless it is on x86 platform, wouldn't it?

fortunately no, because ARM machines won't pass `cpuinfo_has_x86_f16c()`.

swolchok replied:
> also available on ARM

I don't think so? https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mkldnn/Utils.h#L120

something to look at for BF16 though.

jgong5 replied:
> oh, we use FP32 accumulation as well.

Will you remove this `cpuinfo_has_x86_avx512fp16()` check then? I don't think it is relevant.

swolchok replied:
>  remove this cpuinfo_has_x86_avx512fp16() check then? I don't think it is relevant.

I am surprised, but you're definitely the authority on this and I don't have a Sapphire Rapids machine to test on. I'll leave a note for posterity though.

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Testing: To check perf, I ran python torchchat.py generate stories110M
--dtype fp16 --device cpu on an x86 machine without AVX512FP16. Observed roughly 5x tokens/sec increase.
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Hook up fp16_gemv_trans to x86 fp16 GEMM"


This is the first big milestone we've been building towards!
(Following rev also hooks this up to actual gemv.)
Testing: To check perf, I ran python torchchat.py generate stories110M
--dtype fp16 --device cpu on an x86 machine without AVX512FP16. Observed roughly 5x tokens/sec increase.
Differential Revision: [D64280688](https://our.internmc.facebook.com/intern/diff/D64280688/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D64280688/)!

[ghstack-poisoned]
